chapter4/newton-solver.py

```diff
@@ -0,0 +1,300 @@
+
+
+def u_exact(x):
+    return eval(str(u_ufl))
+
+
+# -
+
+# Next, we define the boundary condition `bc`, the residual `F` and the Jacobian `J`.
+
+# +
+V = dolfinx.fem.FunctionSpace(domain, ("CG", 1))
+u_D = dolfinx.fem.Function(V)
+u_D.interpolate(u_exact)
+fdim = domain.topology.dim - 1
+domain.topology.create_connectivity(fdim, fdim + 1)
+boundary_facets = dolfinx.mesh.exterior_facet_indices(domain.topology)
+bc = dolfinx.fem.dirichletbc(u_D, dolfinx.fem.locate_dofs_topological(V, fdim, boundary_facets))
+
+uh = dolfinx.fem.Function(V)
+v = ufl.TestFunction(V)
+F = q(uh) * ufl.dot(ufl.grad(uh), ufl.grad(v)) * ufl.dx - f * v * ufl.dx
+J = ufl.derivative(F, uh)
+residual = dolfinx.fem.form(F)
+jacobian = dolfinx.fem.form(J)
+
+# -
+# Next, we define the matrix `A`, right hand side vector `L` and the correction function `dx`
+
+dx = dolfinx.fem.Function(V)
+A = dolfinx.fem.petsc.create_matrix(jacobian)
+L = dolfinx.fem.petsc.create_vector(residual)
+solver = PETSc.KSP().create(mesh.comm)
+solver.setOperators(A)
+
+# As we for this problem has strong Dirichlet conditions, we need to apply lifting to the right hand side of our Newton problem.
+# We previously had that we wanted to solve the system:
+# $$\begin{align}
+# J_F(x_k)\delta x_k &= - F(x_k)\\
+# x_{k+1} &= x_k + \delta x_k
+# \end{align}$$
+# we want $x_{k+1}\vert_{bc}= u_D$. However, we do not know if $x_k\vert_{bc}=u_D$.
+# Therefore, we want to apply the following boundary condition for our correction $\delta x_k$
+# $$\begin{align}
+# \delta x_k\vert_{bc} = u_D-x_k\vert_{bc}
+# \end{align}$$
+# We therefore arrive at the following Newton scheme
+
+
+i = 0
+error = dolfinx.fem.form(ufl.inner(uh - u_ufl, uh - u_ufl) * ufl.dx(metadata={"quadrature_degree": 4}))
+L2_error = []
+dx_norm = []
+while i < max_iterations:
+    # Assemble Jacobian and residual
+    L.zeroEntries()
+    A.zeroEntries()
+    dolfinx.fem.petsc.assemble_matrix(A, jacobian, bcs=[bc])
+    A.assemble()
+    dolfinx.fem.petsc.assemble_vector(L, residual)
+    L.scale(-1)
+    # Compute b - J(u_D-u_(i-1))
+    dolfinx.fem.petsc.apply_lifting(L, [jacobian], [[bc]], x0=[uh.vector], scale=1) 
+    L.ghostUpdate(addv=PETSc.InsertMode.ADD, mode=PETSc.ScatterMode.REVERSE)
+    # Set dx|_bc = u_{i-1}-u_D
+    dolfinx.fem.petsc.set_bc(L, [bc], uh.vector, 1.0)
+    L.ghostUpdate(addv=PETSc.InsertMode.INSERT_VALUES, mode=PETSc.ScatterMode.FORWARD)
+
+    # Solve linear problem
+    solver.solve(L, dx.vector)
+    dx.x.scatter_forward()
+
+    # Update u_{i+1} = u_i + delta x_i
+    uh.x.array[:] += dx.x.array
+    i += 1
+
+    # Compute norm of update
+    correction_norm = dx.vector.norm(0)
+
+    # Compute L2 error comparing to the analytical solution
+    L2_error.append(np.sqrt(mesh.comm.allreduce(dolfinx.fem.assemble_scalar(error), op=MPI.SUM)))
+    dx_norm.append(correction_norm)
+    
+    print(f"Iteration {i}: Correction norm {correction_norm}, L2 error: {L2_error[-1]}")
+    if correction_norm < 1e-10:
```

Comment:
I am guessing this corresponds to "incremental" as the `convergence_criterion` when using an inbuilt newton solver. 

[Update] Indeed it is!